src/lib.rs

```diff
@@ -1,32 +1,49 @@
 use async_std::fs::{create_dir_all, set_permissions, Permissions};
 use std::{
+    error::Error,
     fs::File,
     io,
     path::{Path, PathBuf},
 };
 use zip::read::ZipFile;
+use zip::result::ZipError;
 
-pub async fn unzip_archive(file: File, directory: &Path) -> Result<(), Box<dyn std::error::Error>> {
+#[derive(Debug)]
+pub struct FileUnzipResult {
+    pub name: Option<String>,
+    pub result: Result<(), UnzipError>,
+}
+
+#[derive(Debug)]
+pub enum UnzipError {
+    OtherError,
+    ZipError(ZipError),
+}
+
+pub async fn unzip_archive(
+    file: File,
+    directory: &Path,
+) -> Result<Vec<FileUnzipResult>, Box<dyn Error>> {
     let mut archive = zip::ZipArchive::new(file)?;
     let mut results = vec![];
 
     for file_index in 0..archive.len() {
         match archive.by_index(file_index) {
```

Comment:
We can already handle some of the errors:
http://mvdnes.github.io/rust-docs/zip-rs/zip/result/enum.ZipError.html
I will take a look today, I think we can already handle some of the exit codes!
Note: ZipError is returned in both `by_index`, but also `ZipArchive::new`, so there is some overlap there we need to handle!